File: agentmesh/events.py
"""AMP v0.3 event model. See docs/amp-spec.md."""

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict

EVENT_TYPES = frozenset(
    {
        "agent_start",
        "agent_end",
        "agent_error",
        "message",
        "tool_call",
    }
)

# Event types that may carry the optional AMP v0.3 usage payload fields.
USAGE_EVENT_TYPES = frozenset({"agent_end", "tool_call"})


def utc_now_iso() -> str:
    """Current time as an ISO-8601 UTC string."""
    return datetime.now(timezone.utc).isoformat(timespec="microseconds")


def _parse_ts(ts: str) -> datetime:
    # Python 3.9's fromisoformat does not accept a trailing "Z".
    if ts.endswith("Z"):
        ts = ts[:-1] + "+00:00"
    return datetime.fromisoformat(ts)


def _validate_usage(payload: Dict[str, Any]) -> None:
    """Validate the optional AMP v0.3 usage payload keys, if present.

    Absent keys are always fine (v0.1/v0.2 events validate unchanged); a
    present key with the wrong type or a negative value raises ValueError.
    """
    for name in ("input_tokens", "output_tokens"):
        if name in payload:
            value = payload[name]
            if not isinstance(value, int) or isinstance(value, bool) or value < 0:
                raise ValueError("%s must be a non-negative integer, got %r" % (name, value))
    if "cost_usd" in payload:
        value = payload["cost_usd"]
        if not isinstance(value, (int, float)) or isinstance(value, bool) or value < 0:
            raise ValueError("cost_usd must be a non-negative number, got %r" % (value,))
    if "model" in payload:
        value = payload["model"]
        if not isinstance(value, str) or not value:
            raise ValueError("model must be a non-empty string, got %r" % (value,))


@dataclass(frozen=True)
class Event:
    """A single AMP v0.3 event."""

    event_id: str
    run_id: str
    agent: str
    type: str
    ts: str
    payload: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.validate()

    def validate(self) -> None:
        """Raise ValueError if this event violates the AMP v0.3 spec."""
        for name in ("event_id", "run_id", "agent"):
            value = getattr(self, name)
            if not isinstance(value, str) or not value:
                raise ValueError("%s must be a non-empty string, got %r" % (name, value))
        if self.type not in EVENT_TYPES:
            raise ValueError("type must be one of %s, got %r" % (sorted(EVENT_TYPES), self.type))
        if not isinstance(self.ts, str):
            raise ValueError("ts must be an ISO-8601 string, got %r" % (self.ts,))
        try:
            parsed = _parse_ts(self.ts)
        except ValueError:
            raise ValueError("ts is not valid ISO-8601: %r" % (self.ts,))
        if parsed.utcoffset() is None or parsed.utcoffset().total_seconds() != 0:
            raise ValueError("ts must be timezone-aware UTC: %r" % (self.ts,))
        if not isinstance(self.payload, dict):
            raise ValueError("payload must be a dict, got %r" % (self.payload,))
        if self.type in USAGE_EVENT_TYPES:
            _validate_usage(self.payload)

    @classmethod
    def new(
        cls,
        run_id: str,
        agent: str,
        type: str,
        payload: Dict[str, Any],
    ) -> "Event":
        """Create a validated event with a fresh event_id and current UTC ts."""
        return cls(
            event_id=str(uuid.uuid4()),
            run_id=run_id,
            agent=agent,
            type=type,
            ts=utc_now_iso(),
            payload=payload,
        )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "event_id": self.event_id,
            "run_id": self.run_id,
            "agent": self.agent,
            "type": self.type,
            "ts": self.ts,
            "payload": self.payload,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Event":
        return cls(
            event_id=data["event_id"],
            run_id=data["run_id"],
            agent=data["agent"],
            type=data["type"],
            ts=data["ts"],
            payload=data["payload"],
        )

File: sdk-ts/src/events.ts
/**
 * AMP v0.2 event model — TypeScript mirror of agentmesh/events.py.
 * See docs/amp-spec.md in the repository root.
 */

import { randomUUID } from "node:crypto";

export const EVENT_TYPES = [
  "agent_start",
  "agent_end",
  "agent_error",
  "message",
  "tool_call",
] as const;

export type EventType = (typeof EVENT_TYPES)[number];

/** A single AMP v0.2 event (see docs/amp-spec.md). */
export interface AmpEvent {
  event_id: string;
  run_id: string;
  agent: string;
  type: EventType;
  ts: string;
  payload: Record<string, unknown>;
}

/** Current time as an ISO-8601 UTC string (trailing "Z", valid per AMP). */
export function utcNowIso(): string {
  return new Date().toISOString();
}

/** Create a validated event with a fresh event_id and current UTC ts. */
export function newEvent(
  runId: string,
  agent: string,
  type: EventType,
  payload: Record<string, unknown>,
): AmpEvent {
  return validateEvent({
    event_id: randomUUID(),
    run_id: runId,
    agent,
    type,
    ts: utcNowIso(),
    payload,
  });
}

// ISO-8601 timestamp with an explicit timezone, e.g.
// 2026-06-10T08:15:30.123456+00:00 or 2026-06-10T08:15:30.123Z.
const ISO_8601_RE = /^\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$/;

function isUtcIso8601(ts: string): boolean {
  const match = ISO_8601_RE.exec(ts);
  if (match === null) {
    return false;
  }
  if (Number.isNaN(Date.parse(ts.replace(" ", "T")))) {
    return false;
  }
  // Mirror Python: timezone-aware with a UTC offset of zero.
  return ts.endsWith("Z") || ts.endsWith("+00:00") || ts.endsWith("-00:00");
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Validate one AMP v0.2 event, mirroring Event.validate() in
 * agentmesh/events.py. Throws an Error describing the first violation;
 * returns the event (narrowed to AmpEvent) when valid.
 *
 * Rules: all six fields present; event_id/run_id/agent non-empty strings;
 * type one of EVENT_TYPES; ts ISO-8601 with a UTC offset of zero; payload
 * a JSON object. v0.2 span fields (payload.span_id / payload.parent_span_id)
 * are optional but must be non-empty strings when present.
 */
export function validateEvent(event: unknown): AmpEvent {
  if (!isPlainObject(event)) {
    throw new Error(`event must be an object, got ${JSON.stringify(event)}`);
  }
  for (const name of ["event_id", "run_id", "agent"] as const) {
    const value = event[name];
    if (typeof value !== "string" || value === "") {
      throw new Error(`${name} must be a non-empty string, got ${JSON.stringify(value)}`);
    }
  }
  const type = event["type"];
  if (typeof type !== "string" || !(EVENT_TYPES as readonly string[]).includes(type)) {
    throw new Error(
      `type must be one of ${JSON.stringify(EVENT_TYPES)}, got ${JSON.stringify(type)}`,
    );
  }
  const ts = event["ts"];
  if (typeof ts !== "string" || !isUtcIso8601(ts)) {
    throw new Error(`ts must be an ISO-8601 UTC string, got ${JSON.stringify(ts)}`);
  }
  const payload = event["payload"];
  if (!isPlainObject(payload)) {
    throw new Error(`payload must be an object, got ${JSON.stringify(payload)}`);
  }
  for (const name of ["span_id", "parent_span_id"] as const) {
    if (name in payload) {
      const value = payload[name];
      if (typeof value !== "string" || value === "") {
        throw new Error(
          `payload.${name} must be a non-empty string when present, got ${JSON.stringify(value)}`,
        );
      }
    }
  }
  return event as unknown as AmpEvent;
}
